### PR TITLE
mfem and petsc: alter hypre dependencies

### PR DESCRIPTION
**Pull request by cessenat**
Hypre latest version 2.22.1 breaks MFEM and PETSc.
This PR is suggested by @scheibelp for MFEM and @balay for PETSc.
This does alter sundial, despite @balay reports it may be broken by Hypre update from PR   https://github.com/spack/spack/pull/25781
To be complete, I changed a bit what @scheibelp advised since MFEM 4.3 accepts hypre up to 2.22.0 not only 2.20.0.

### CODE REVIEW

File: var/spack/repos/builtin/packages/mfem/package.py [hunk 1]
```diff
@@ -200,7 +200,7 @@ class Mfem(Package, CudaPackage, ROCmPackage):
     depends_on('mpi', when='+mpi')
     depends_on('hypre@2.10.0:2.13.99', when='@:3.3.99+mpi')
     depends_on('hypre@:2.20.0', when='@3.4:4.2.99+mpi')
-    depends_on('hypre', when='@4.3.0:+mpi')
+    depends_on('hypre@:2.22.0', when='@4.3.0:+mpi')
```
Comment (v-dobrev) [inline var/spack/repos/builtin/packages/mfem/package.py]: This should be:
```python
    depends_on('hypre@:2.22.0', when='@4.3.0+mpi')
    depends_on('hypre', when='+mpi')
```
In particular, `mfem@develop` should work with new versions of hypre.

**Comment by v-dobrev** [comment]
For the upcoming xSDK v0.7.0, we may actually add a patch for mfem v4.3 to make sure it works with hypre v2.22.1.

@balay, let me know if this will be necessary for xSDK v0.7.0. The only other option I see is to use hypre v2.22.0 instead of v2.22.1.

**Comment by balay** [comment]
for xsdk@0.7.0 - I see hypre will be at 2.23.0. Since mfem will be at 4.3.0 - a patch (in spack) would be needed - unless there will be a patch update release for mfem [with this fix]

File: var/spack/repos/builtin/packages/mfem/package.py (hunk 1 above)
Comment (cessenat) [reply var/spack/repos/builtin/packages/mfem/package.py]: That is right. I make a commit for that, thanks.

**Comment by cessenat** [comment]
Trying to rerun the pipeline

File: var/spack/repos/builtin/packages/mfem/package.py (hunk 1 above)
Comment (scheibelp) [reply var/spack/repos/builtin/packages/mfem/package.py]: @cessenat it looks like this still needs to be updated

Comment (cessenat) [reply var/spack/repos/builtin/packages/mfem/package.py]: Sorry @scheibelp but I altered your advice from 

```
    depends_on('hypre@2.10.0:2.13.99', when='@:3.3.99+mpi')
    depends_on('hypre@:2.20.0', when='@3.4:4.3.0+mpi')
```
to
 ```
    depends_on('hypre@2.10.0:2.13.99', when='@:3.3.99+mpi')
    depends_on('hypre@:2.20.0', when='@3.4:4.2.99+mpi')
    depends_on('hypre@:2.22.0', when='@4.3.0:+mpi')
    depends_on('hypre', when='+mpi')
```
Tell me if I made another mistake maybe misusing git...

Comment (cessenat) [reply var/spack/repos/builtin/packages/mfem/package.py]: Oh very sorry @scheibelp you were right, 4.3.0:+mpi was wrong, it had to be 4.3.0+mpi.
Corrected at last, and apologize for the waste of time.

**Review by v-dobrev — APPROVED** [review]

**Review by tzanio — APPROVED** [review]

**Review by scheibelp — APPROVED** [review]

File: var/spack/repos/builtin/packages/mfem/package.py (hunk 1 above)
Comment (scheibelp) [reply var/spack/repos/builtin/packages/mfem/package.py]: No problem these things can be easy to miss. Thanks for updating it!

**Comment by cessenat** [comment]
Thanks everybody

**Pull request merged**